{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Jupyter Notebook Tutorial for Anthropology Course\n",
    "\n",
    "## Exploring 1000 Genomes Data using Basic Python\n",
    "\n",
    "### Introduction\n",
    "We will explore genetic variation, allele frequencies, and the Hardy-Weinberg equilibrium using data from the 1000 Genomes Project.\n",
    "\n",
    "### Table of Contents\n",
    "1. [Environment Setup](#Environment-Setup)\n",
    "2. [Data Import](#Data-Import)\n",
    "3. [Data Exploration](#Data-Exploration)\n",
    "4. [Genetic Variation Analysis](#Genetic-Variation-Analysis)\n",
    "5. [Allele Frequencies](#Allele-Frequencies)\n",
    "6. [Hardy-Weinberg Equilibrium](#Hardy-Weinberg-Equilibrium)\n",
    "7. [Conclusion](#Conclusion)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Environment-Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import importlib.util\n",
    "\n",
    "# Function to check if a package is installed\n",
    "def check_install_package(package_name, pip_name=None):\n",
    "    package_spec = importlib.util.find_spec(package_name)\n",
    "    if package_spec is None:\n",
    "        print(f\"{package_name} not found. Installing...\")\n",
    "        !pip install {pip_name if pip_name else package_name}\n",
    "    else:\n",
    "        print(f\"{package_name} is already installed.\")\n",
    "\n",
    "# Check and install scikit-allel if not installed\n",
    "check_install_package('allel', 'scikit-allel')\n",
    "check_install_package('zarr', 'zarr')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/lakishadavid/anthropology_methods_course'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os # Imports Python's built-in os library, which allows us to interact with the operating system.\n",
    "primary_directory = os.getcwd() # get the current working directory and stores it in the variable primary_directory\n",
    "primary_directory # Displays the value of primary_directory"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you want to change the value of primary_directory to a different directory, you can manually set it like this:\n",
    "\n",
    "`primary_directory = \"/path/to/your/desired/directory\"`\n",
    "\n",
    "`primary_directory = \"/home/username/MyProject\"`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/lakishadavid'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# To change primary_directory, replace the path below to the path you want your for your files\n",
    "primary_directory = '/home/lakishadavid' # for example, this is for my environment\n",
    "primary_directory"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The followng code checks if the directories already exist before attempting to create them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Directories have been set up!\n"
     ]
    }
   ],
   "source": [
    "# Define the other directories\n",
    "use_directory = os.path.join(primary_directory, \"use\")\n",
    "results_directory = os.path.join(primary_directory, \"results\")\n",
    "references_directory = os.path.join(primary_directory, \"references\")\n",
    "data_directory = os.path.join(primary_directory, \"data\")\n",
    "\n",
    "# Create the directories if they do not exist\n",
    "for directory in [use_directory, results_directory, references_directory, data_directory]:\n",
    "    if not os.path.exists(directory):\n",
    "        os.makedirs(directory)\n",
    "\n",
    "print(\"Directories have been set up!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The line `if not os.path.exists(directory):` checks whether a directory with the specified path already exists. If it does not exist, the code inside the if block, `os.makedirs(directory)`, will execute, creating the directory. If the directory already exists, the `os.makedirs(directory)` line will not be executed for that directory, and the code will simply move on to the next directory in the list to check. No new directories will be created if they already exist, and no existing data or files within those directories will be affected."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Import"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Introduction to the IGSR 30x GRCh38 Data Collection\n",
    "The International Genome Sample Resource (IGSR) provides a data collection for the 30x GRCh38 human genome assembly. This resource is invaluable for researchers and scientists who are working on genomics, as it offers high-quality, publicly available data sets. The 30x coverage ensures a high level of accuracy and reliability for genomic studies."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## How to Download Populations File and Sample File\n",
    "\n",
    "Normally, you would find options to download various data files, including the populations file and the sample file, on the IGSR data portal. Here's a general outline of how you could download these files and save them to your references_directory:\n",
    "\n",
    "Navigate to the Data Download Section: Go to the IGSR data portal at https://www.internationalgenome.org/data-portal/data-collection/30x-grch38. Locate the buttons that says \"Download the list\" in the sample section and the population section.\n",
    "\n",
    "Download Files: Click on the download links for the populations file and the sample file. Note the directory where these files are saved and the filenames.\n",
    "\n",
    "Move Files to references_directory: Move these downloaded files to your reference directory. Run the cell below to remind yourself of how to navigate to your reference directory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/lakishadavid/genetic_genealogy/references\n"
     ]
    }
   ],
   "source": [
    "# Your reference directory\n",
    "print(references_directory)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## File Verification\n",
    "\n",
    "Before proceeding with the data subsetting, let's ensure that the sample and population files you intend to use are available in the `references_directory`.\n",
    "\n",
    "If you are using VSCode to run this Jupyter Notebook, when you run the next cell, a popup box should appear asking you to enter the filenames for the sample and population files. Please enter the filenames in these boxes to proceed.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using sample file: onekg_grch38_samples.tsv\n",
      "Using population file: onekg_grch38_populations.tsv\n",
      "The sample file /home/lakishadavid/genetic_genealogy/references/onekg_grch38_samples.tsv exists.\n",
      "The population file /home/lakishadavid/genetic_genealogy/references/onekg_grch38_populations.tsv exists.\n"
     ]
    }
   ],
   "source": [
    "sample_file_name = input(\"Please enter the name of your sample file (e.g., onekg_grch38_samples.tsv) or press ENTER for default: \")\n",
    "if not sample_file_name:\n",
    "    sample_file_name = \"onekg_grch38_samples.tsv\"\n",
    "print(f\"Using sample file: {sample_file_name}\")\n",
    "\n",
    "population_file_name = input(\"Please enter the name of your population file (e.g., onekg_grch38_populations.tsv) or press ENTER for default: \")\n",
    "if not population_file_name:\n",
    "    population_file_name = \"onekg_grch38_populations.tsv\"\n",
    "print(f\"Using population file: {population_file_name}\")\n",
    "\n",
    "# Check if the files exist in the references_directory\n",
    "sample_file_path = os.path.join(references_directory, sample_file_name)\n",
    "population_file_path = os.path.join(references_directory, population_file_name)\n",
    "\n",
    "# Check if sample file exists\n",
    "if os.path.exists(sample_file_path):\n",
    "    print(f\"The sample file {sample_file_path} exists.\")\n",
    "else:\n",
    "    print(f\"The sample file {sample_file_path} does not exist.\")\n",
    "\n",
    "# Check if population file exists\n",
    "if os.path.exists(population_file_path):\n",
    "    print(f\"The population file {population_file_path} exists.\")\n",
    "else:\n",
    "    print(f\"The population file {population_file_path} does not exist.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exploring the Sample and Population Files\n",
    "\n",
    "Before diving into the analysis, it's a good idea to explore the sample and population files to get a sense of what the data looks like. We'll use Pandas to open these files and display the first few rows.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First few rows of the sample file:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample name</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Biosample ID</th>\n",
       "      <th>Population code</th>\n",
       "      <th>Population name</th>\n",
       "      <th>Superpopulation code</th>\n",
       "      <th>Superpopulation name</th>\n",
       "      <th>Population elastic ID</th>\n",
       "      <th>Data collections</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HG00271</td>\n",
       "      <td>male</td>\n",
       "      <td>SAME123417</td>\n",
       "      <td>FIN</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European Ancestry</td>\n",
       "      <td>FIN</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HG00276</td>\n",
       "      <td>female</td>\n",
       "      <td>SAME123424</td>\n",
       "      <td>FIN</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European Ancestry</td>\n",
       "      <td>FIN</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HG00288</td>\n",
       "      <td>female</td>\n",
       "      <td>SAME1839246</td>\n",
       "      <td>FIN</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European Ancestry</td>\n",
       "      <td>FIN</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HG00290</td>\n",
       "      <td>male</td>\n",
       "      <td>SAME1839057</td>\n",
       "      <td>FIN</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European Ancestry</td>\n",
       "      <td>FIN</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HG00303</td>\n",
       "      <td>male</td>\n",
       "      <td>SAME1840115</td>\n",
       "      <td>FIN</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European Ancestry</td>\n",
       "      <td>FIN</td>\n",
       "      <td>1000 Genomes on GRCh38</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sample name     Sex Biosample ID Population code Population name  \\\n",
       "0     HG00271    male   SAME123417             FIN         Finnish   \n",
       "1     HG00276  female   SAME123424             FIN         Finnish   \n",
       "2     HG00288  female  SAME1839246             FIN         Finnish   \n",
       "3     HG00290    male  SAME1839057             FIN         Finnish   \n",
       "4     HG00303    male  SAME1840115             FIN         Finnish   \n",
       "\n",
       "  Superpopulation code Superpopulation name Population elastic ID  \\\n",
       "0                  EUR    European Ancestry                   FIN   \n",
       "1                  EUR    European Ancestry                   FIN   \n",
       "2                  EUR    European Ancestry                   FIN   \n",
       "3                  EUR    European Ancestry                   FIN   \n",
       "4                  EUR    European Ancestry                   FIN   \n",
       "\n",
       "                                    Data collections  \n",
       "0  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "1  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "2  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "3  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "4                             1000 Genomes on GRCh38  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First few rows of the population file:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Population code</th>\n",
       "      <th>Population elastic ID</th>\n",
       "      <th>Population name</th>\n",
       "      <th>Population description</th>\n",
       "      <th>Population latitude</th>\n",
       "      <th>Population longitude</th>\n",
       "      <th>Superpopulation code</th>\n",
       "      <th>Superpopulation name</th>\n",
       "      <th>Superpopulation display colour</th>\n",
       "      <th>Superpopulation display order</th>\n",
       "      <th>Data collections</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FIN</td>\n",
       "      <td>FIN</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>Finnish in Finland</td>\n",
       "      <td>60.17000</td>\n",
       "      <td>24.930000</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European Ancestry</td>\n",
       "      <td>#018ead</td>\n",
       "      <td>4</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS</td>\n",
       "      <td>CHS</td>\n",
       "      <td>Southern Han Chinese</td>\n",
       "      <td>Han Chinese South</td>\n",
       "      <td>23.13333</td>\n",
       "      <td>113.266667</td>\n",
       "      <td>EAS</td>\n",
       "      <td>East Asian Ancestry</td>\n",
       "      <td>#778500</td>\n",
       "      <td>3</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KHV</td>\n",
       "      <td>KHV</td>\n",
       "      <td>Kinh Vietnamese</td>\n",
       "      <td>Kinh in Ho Chi Minh City, Vietnam</td>\n",
       "      <td>10.78000</td>\n",
       "      <td>106.680000</td>\n",
       "      <td>EAS</td>\n",
       "      <td>East Asian Ancestry</td>\n",
       "      <td>#778500</td>\n",
       "      <td>3</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BEB</td>\n",
       "      <td>BEB</td>\n",
       "      <td>Bengali</td>\n",
       "      <td>Bengali in Bangladesh</td>\n",
       "      <td>23.70000</td>\n",
       "      <td>90.350000</td>\n",
       "      <td>SAS</td>\n",
       "      <td>South Asian Ancestry</td>\n",
       "      <td>#c44cfd</td>\n",
       "      <td>5</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PUR</td>\n",
       "      <td>PUR</td>\n",
       "      <td>Puerto Rican</td>\n",
       "      <td>Puerto Rican in Puerto Rico</td>\n",
       "      <td>18.40000</td>\n",
       "      <td>-66.100000</td>\n",
       "      <td>AMR</td>\n",
       "      <td>American Ancestry</td>\n",
       "      <td>#710027</td>\n",
       "      <td>2</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Population code Population elastic ID       Population name  \\\n",
       "0             FIN                   FIN               Finnish   \n",
       "1             CHS                   CHS  Southern Han Chinese   \n",
       "2             KHV                   KHV       Kinh Vietnamese   \n",
       "3             BEB                   BEB               Bengali   \n",
       "4             PUR                   PUR          Puerto Rican   \n",
       "\n",
       "              Population description  Population latitude  \\\n",
       "0                 Finnish in Finland             60.17000   \n",
       "1                  Han Chinese South             23.13333   \n",
       "2  Kinh in Ho Chi Minh City, Vietnam             10.78000   \n",
       "3              Bengali in Bangladesh             23.70000   \n",
       "4        Puerto Rican in Puerto Rico             18.40000   \n",
       "\n",
       "   Population longitude Superpopulation code  Superpopulation name  \\\n",
       "0             24.930000                  EUR     European Ancestry   \n",
       "1            113.266667                  EAS   East Asian Ancestry   \n",
       "2            106.680000                  EAS   East Asian Ancestry   \n",
       "3             90.350000                  SAS  South Asian Ancestry   \n",
       "4            -66.100000                  AMR     American Ancestry   \n",
       "\n",
       "  Superpopulation display colour  Superpopulation display order  \\\n",
       "0                        #018ead                              4   \n",
       "1                        #778500                              3   \n",
       "2                        #778500                              3   \n",
       "3                        #c44cfd                              5   \n",
       "4                        #710027                              2   \n",
       "\n",
       "                                    Data collections  \n",
       "0  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "1  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "2  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "3  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "4  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Import the Pandas library\n",
    "import pandas as pd\n",
    "\n",
    "# Load the sample and population files into Pandas DataFrames\n",
    "try:\n",
    "    sample_df = pd.read_csv(sample_file_path, sep='\\t')\n",
    "    population_df = pd.read_csv(population_file_path, sep='\\t')\n",
    "    \n",
    "    # Display the first few rows of each DataFrame\n",
    "    print(\"First few rows of the sample file:\")\n",
    "    display(sample_df.head())\n",
    "    \n",
    "    print(\"First few rows of the population file:\")\n",
    "    display(population_df.head())\n",
    "    \n",
    "except FileNotFoundError as e:\n",
    "    print(f\"File not found: {e}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Files Overview\n",
    "\n",
    "Before diving into the analysis, it's crucial to understand the data files we'll be working with.\n",
    "\n",
    "---\n",
    "\n",
    "### Populations File\n",
    "\n",
    "#### Description\n",
    "The populations file contains information about the various populations that are part of the genomic study. This file is essential for understanding the diversity of the samples and for performing population-specific analyses.\n",
    "\n",
    "#### Typical Columns\n",
    "- **Population ID**: Unique identifier for each population.\n",
    "- **Population Name**: Name of the population.\n",
    "- **Region**: Geographical region where the population is located.\n",
    "- **Number of Samples**: Number of samples collected from this population.\n",
    "- **Other Metadata**: Additional information such as ethnicity, age range, etc.\n",
    "\n",
    "#### Use Cases\n",
    "- Filtering genomic data based on specific populations.\n",
    "- Performing population-specific genetic variation analyses.\n",
    "- Understanding the distribution of samples across different populations.\n",
    "\n",
    "---\n",
    "\n",
    "### Sample File\n",
    "\n",
    "#### Description\n",
    "The sample file contains detailed information about each individual sample that is part of the study. This file is essential for tracking the source of each genomic sequence and for associating it with specific traits or conditions.\n",
    "\n",
    "#### Typical Columns\n",
    "- **Sample ID**: Unique identifier for each sample.\n",
    "- **Population ID**: The population to which the sample belongs.\n",
    "- **Gender**: Gender of the individual from whom the sample was taken.\n",
    "- **Age**: Age of the individual.\n",
    "- **Health Status**: Information about the health condition of the individual, if applicable.\n",
    "- **Other Metadata**: Additional information such as the date of sample collection, sequencing technology used, etc.\n",
    "\n",
    "#### Use Cases\n",
    "- Filtering genomic data based on specific samples or traits.\n",
    "- Performing individual-level analyses.\n",
    "- Associating genomic variations with specific traits or conditions.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Exploratory Data Analysis with Pandas\n",
    "\n",
    "## Introduction\n",
    "Before diving into more complex analyses, it's essential to understand the structure and characteristics of your data. Pandas is a powerful Python library that provides fast, flexible, and expressive data structures designed to make working with \"relational\" or \"labeled\" data both easy and intuitive. Let's explore some basic Pandas functionalities to better understand our sample and population files.\n",
    "\n",
    "Remember that we loaded the sample and population files earlier and created Pandas Dataframes called sample_df and population_df.\n",
    "\n",
    "### Basic Information\n",
    "You can get a quick overview of the DataFrame using .info()."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 2709 entries, 0 to 2708\n",
      "Data columns (total 9 columns):\n",
      " #   Column                 Non-Null Count  Dtype \n",
      "---  ------                 --------------  ----- \n",
      " 0   Sample name            2709 non-null   object\n",
      " 1   Sex                    2709 non-null   object\n",
      " 2   Biosample ID           2709 non-null   object\n",
      " 3   Population code        2709 non-null   object\n",
      " 4   Population name        2709 non-null   object\n",
      " 5   Superpopulation code   2709 non-null   object\n",
      " 6   Superpopulation name   2709 non-null   object\n",
      " 7   Population elastic ID  2709 non-null   object\n",
      " 8   Data collections       2709 non-null   object\n",
      "dtypes: object(9)\n",
      "memory usage: 190.6+ KB\n"
     ]
    }
   ],
   "source": [
    "# Get basic information about the sample DataFrame\n",
    "sample_df.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can preview the dataframe by viewing the first rows using .head() or the last rows using .tail, default 5."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample name</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Biosample ID</th>\n",
       "      <th>Population code</th>\n",
       "      <th>Population name</th>\n",
       "      <th>Superpopulation code</th>\n",
       "      <th>Superpopulation name</th>\n",
       "      <th>Population elastic ID</th>\n",
       "      <th>Data collections</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HG00271</td>\n",
       "      <td>male</td>\n",
       "      <td>SAME123417</td>\n",
       "      <td>FIN</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European Ancestry</td>\n",
       "      <td>FIN</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HG00276</td>\n",
       "      <td>female</td>\n",
       "      <td>SAME123424</td>\n",
       "      <td>FIN</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European Ancestry</td>\n",
       "      <td>FIN</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HG00288</td>\n",
       "      <td>female</td>\n",
       "      <td>SAME1839246</td>\n",
       "      <td>FIN</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European Ancestry</td>\n",
       "      <td>FIN</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HG00290</td>\n",
       "      <td>male</td>\n",
       "      <td>SAME1839057</td>\n",
       "      <td>FIN</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European Ancestry</td>\n",
       "      <td>FIN</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HG00303</td>\n",
       "      <td>male</td>\n",
       "      <td>SAME1840115</td>\n",
       "      <td>FIN</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European Ancestry</td>\n",
       "      <td>FIN</td>\n",
       "      <td>1000 Genomes on GRCh38</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sample name     Sex Biosample ID Population code Population name  \\\n",
       "0     HG00271    male   SAME123417             FIN         Finnish   \n",
       "1     HG00276  female   SAME123424             FIN         Finnish   \n",
       "2     HG00288  female  SAME1839246             FIN         Finnish   \n",
       "3     HG00290    male  SAME1839057             FIN         Finnish   \n",
       "4     HG00303    male  SAME1840115             FIN         Finnish   \n",
       "\n",
       "  Superpopulation code Superpopulation name Population elastic ID  \\\n",
       "0                  EUR    European Ancestry                   FIN   \n",
       "1                  EUR    European Ancestry                   FIN   \n",
       "2                  EUR    European Ancestry                   FIN   \n",
       "3                  EUR    European Ancestry                   FIN   \n",
       "4                  EUR    European Ancestry                   FIN   \n",
       "\n",
       "                                    Data collections  \n",
       "0  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "1  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "2  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "3  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "4                             1000 Genomes on GRCh38  "
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample name</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Biosample ID</th>\n",
       "      <th>Population code</th>\n",
       "      <th>Population name</th>\n",
       "      <th>Superpopulation code</th>\n",
       "      <th>Superpopulation name</th>\n",
       "      <th>Population elastic ID</th>\n",
       "      <th>Data collections</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HG00271</td>\n",
       "      <td>male</td>\n",
       "      <td>SAME123417</td>\n",
       "      <td>FIN</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European Ancestry</td>\n",
       "      <td>FIN</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HG00276</td>\n",
       "      <td>female</td>\n",
       "      <td>SAME123424</td>\n",
       "      <td>FIN</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European Ancestry</td>\n",
       "      <td>FIN</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HG00288</td>\n",
       "      <td>female</td>\n",
       "      <td>SAME1839246</td>\n",
       "      <td>FIN</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European Ancestry</td>\n",
       "      <td>FIN</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HG00290</td>\n",
       "      <td>male</td>\n",
       "      <td>SAME1839057</td>\n",
       "      <td>FIN</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European Ancestry</td>\n",
       "      <td>FIN</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HG00303</td>\n",
       "      <td>male</td>\n",
       "      <td>SAME1840115</td>\n",
       "      <td>FIN</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European Ancestry</td>\n",
       "      <td>FIN</td>\n",
       "      <td>1000 Genomes on GRCh38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HG00308</td>\n",
       "      <td>male</td>\n",
       "      <td>SAME124161</td>\n",
       "      <td>FIN</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European Ancestry</td>\n",
       "      <td>FIN</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HG00310</td>\n",
       "      <td>male</td>\n",
       "      <td>SAME124338</td>\n",
       "      <td>FIN</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European Ancestry</td>\n",
       "      <td>FIN</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HG00315</td>\n",
       "      <td>female</td>\n",
       "      <td>SAME124335</td>\n",
       "      <td>FIN</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European Ancestry</td>\n",
       "      <td>FIN</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HG00327</td>\n",
       "      <td>female</td>\n",
       "      <td>SAME123791</td>\n",
       "      <td>FIN</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European Ancestry</td>\n",
       "      <td>FIN</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HG00334</td>\n",
       "      <td>female</td>\n",
       "      <td>SAME123981</td>\n",
       "      <td>FIN</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European Ancestry</td>\n",
       "      <td>FIN</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sample name     Sex Biosample ID Population code Population name  \\\n",
       "0     HG00271    male   SAME123417             FIN         Finnish   \n",
       "1     HG00276  female   SAME123424             FIN         Finnish   \n",
       "2     HG00288  female  SAME1839246             FIN         Finnish   \n",
       "3     HG00290    male  SAME1839057             FIN         Finnish   \n",
       "4     HG00303    male  SAME1840115             FIN         Finnish   \n",
       "5     HG00308    male   SAME124161             FIN         Finnish   \n",
       "6     HG00310    male   SAME124338             FIN         Finnish   \n",
       "7     HG00315  female   SAME124335             FIN         Finnish   \n",
       "8     HG00327  female   SAME123791             FIN         Finnish   \n",
       "9     HG00334  female   SAME123981             FIN         Finnish   \n",
       "\n",
       "  Superpopulation code Superpopulation name Population elastic ID  \\\n",
       "0                  EUR    European Ancestry                   FIN   \n",
       "1                  EUR    European Ancestry                   FIN   \n",
       "2                  EUR    European Ancestry                   FIN   \n",
       "3                  EUR    European Ancestry                   FIN   \n",
       "4                  EUR    European Ancestry                   FIN   \n",
       "5                  EUR    European Ancestry                   FIN   \n",
       "6                  EUR    European Ancestry                   FIN   \n",
       "7                  EUR    European Ancestry                   FIN   \n",
       "8                  EUR    European Ancestry                   FIN   \n",
       "9                  EUR    European Ancestry                   FIN   \n",
       "\n",
       "                                    Data collections  \n",
       "0  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "1  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "2  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "3  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "4                             1000 Genomes on GRCh38  \n",
       "5  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "6  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "7  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "8  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "9  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample name</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Biosample ID</th>\n",
       "      <th>Population code</th>\n",
       "      <th>Population name</th>\n",
       "      <th>Superpopulation code</th>\n",
       "      <th>Superpopulation name</th>\n",
       "      <th>Population elastic ID</th>\n",
       "      <th>Data collections</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2704</th>\n",
       "      <td>NA21143</td>\n",
       "      <td>female</td>\n",
       "      <td>SAME123359</td>\n",
       "      <td>GIH</td>\n",
       "      <td>Gujarati</td>\n",
       "      <td>SAS</td>\n",
       "      <td>South Asian Ancestry</td>\n",
       "      <td>GIH</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2705</th>\n",
       "      <td>NA21112</td>\n",
       "      <td>male</td>\n",
       "      <td>SAME124182</td>\n",
       "      <td>GIH</td>\n",
       "      <td>Gujarati</td>\n",
       "      <td>SAS</td>\n",
       "      <td>South Asian Ancestry</td>\n",
       "      <td>GIH</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2706</th>\n",
       "      <td>NA21117</td>\n",
       "      <td>male</td>\n",
       "      <td>SAME124185</td>\n",
       "      <td>GIH</td>\n",
       "      <td>Gujarati</td>\n",
       "      <td>SAS</td>\n",
       "      <td>South Asian Ancestry</td>\n",
       "      <td>GIH</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2707</th>\n",
       "      <td>NA21124</td>\n",
       "      <td>male</td>\n",
       "      <td>SAME124024</td>\n",
       "      <td>GIH</td>\n",
       "      <td>Gujarati</td>\n",
       "      <td>SAS</td>\n",
       "      <td>South Asian Ancestry</td>\n",
       "      <td>GIH</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2708</th>\n",
       "      <td>NA21129</td>\n",
       "      <td>male</td>\n",
       "      <td>SAME124023</td>\n",
       "      <td>GIH</td>\n",
       "      <td>Gujarati</td>\n",
       "      <td>SAS</td>\n",
       "      <td>South Asian Ancestry</td>\n",
       "      <td>GIH</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Sample name     Sex Biosample ID Population code Population name  \\\n",
       "2704     NA21143  female   SAME123359             GIH        Gujarati   \n",
       "2705     NA21112    male   SAME124182             GIH        Gujarati   \n",
       "2706     NA21117    male   SAME124185             GIH        Gujarati   \n",
       "2707     NA21124    male   SAME124024             GIH        Gujarati   \n",
       "2708     NA21129    male   SAME124023             GIH        Gujarati   \n",
       "\n",
       "     Superpopulation code  Superpopulation name Population elastic ID  \\\n",
       "2704                  SAS  South Asian Ancestry                   GIH   \n",
       "2705                  SAS  South Asian Ancestry                   GIH   \n",
       "2706                  SAS  South Asian Ancestry                   GIH   \n",
       "2707                  SAS  South Asian Ancestry                   GIH   \n",
       "2708                  SAS  South Asian Ancestry                   GIH   \n",
       "\n",
       "                                       Data collections  \n",
       "2704  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "2705  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "2706  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "2707  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "2708  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  "
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample name</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Biosample ID</th>\n",
       "      <th>Population code</th>\n",
       "      <th>Population name</th>\n",
       "      <th>Superpopulation code</th>\n",
       "      <th>Superpopulation name</th>\n",
       "      <th>Population elastic ID</th>\n",
       "      <th>Data collections</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2694</th>\n",
       "      <td>NA20897</td>\n",
       "      <td>male</td>\n",
       "      <td>SAME123664</td>\n",
       "      <td>GIH</td>\n",
       "      <td>Gujarati</td>\n",
       "      <td>SAS</td>\n",
       "      <td>South Asian Ancestry</td>\n",
       "      <td>GIH</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2695</th>\n",
       "      <td>NA20900</td>\n",
       "      <td>female</td>\n",
       "      <td>SAME122920</td>\n",
       "      <td>GIH</td>\n",
       "      <td>Gujarati</td>\n",
       "      <td>SAS</td>\n",
       "      <td>South Asian Ancestry</td>\n",
       "      <td>GIH</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2696</th>\n",
       "      <td>NA20905</td>\n",
       "      <td>male</td>\n",
       "      <td>SAME122915</td>\n",
       "      <td>GIH</td>\n",
       "      <td>Gujarati</td>\n",
       "      <td>SAS</td>\n",
       "      <td>South Asian Ancestry</td>\n",
       "      <td>GIH</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2697</th>\n",
       "      <td>NA21092</td>\n",
       "      <td>male</td>\n",
       "      <td>SAME123322</td>\n",
       "      <td>GIH</td>\n",
       "      <td>Gujarati</td>\n",
       "      <td>SAS</td>\n",
       "      <td>South Asian Ancestry</td>\n",
       "      <td>GIH</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2698</th>\n",
       "      <td>NA21097</td>\n",
       "      <td>female</td>\n",
       "      <td>SAME123319</td>\n",
       "      <td>GIH</td>\n",
       "      <td>Gujarati</td>\n",
       "      <td>SAS</td>\n",
       "      <td>South Asian Ancestry</td>\n",
       "      <td>GIH</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2699</th>\n",
       "      <td>NA21100</td>\n",
       "      <td>male</td>\n",
       "      <td>SAME124440</td>\n",
       "      <td>GIH</td>\n",
       "      <td>Gujarati</td>\n",
       "      <td>SAS</td>\n",
       "      <td>South Asian Ancestry</td>\n",
       "      <td>GIH</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2700</th>\n",
       "      <td>NA21105</td>\n",
       "      <td>male</td>\n",
       "      <td>SAME124437</td>\n",
       "      <td>GIH</td>\n",
       "      <td>Gujarati</td>\n",
       "      <td>SAS</td>\n",
       "      <td>South Asian Ancestry</td>\n",
       "      <td>GIH</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2701</th>\n",
       "      <td>NA20873</td>\n",
       "      <td>male</td>\n",
       "      <td>SAME122841</td>\n",
       "      <td>GIH</td>\n",
       "      <td>Gujarati</td>\n",
       "      <td>SAS</td>\n",
       "      <td>South Asian Ancestry</td>\n",
       "      <td>GIH</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes phase 3 re...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2702</th>\n",
       "      <td>NA20878</td>\n",
       "      <td>female</td>\n",
       "      <td>SAME123595</td>\n",
       "      <td>GIH</td>\n",
       "      <td>Gujarati</td>\n",
       "      <td>SAS</td>\n",
       "      <td>South Asian Ancestry</td>\n",
       "      <td>GIH</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2703</th>\n",
       "      <td>NA20885</td>\n",
       "      <td>male</td>\n",
       "      <td>SAME124525</td>\n",
       "      <td>GIH</td>\n",
       "      <td>Gujarati</td>\n",
       "      <td>SAS</td>\n",
       "      <td>South Asian Ancestry</td>\n",
       "      <td>GIH</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2704</th>\n",
       "      <td>NA21143</td>\n",
       "      <td>female</td>\n",
       "      <td>SAME123359</td>\n",
       "      <td>GIH</td>\n",
       "      <td>Gujarati</td>\n",
       "      <td>SAS</td>\n",
       "      <td>South Asian Ancestry</td>\n",
       "      <td>GIH</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2705</th>\n",
       "      <td>NA21112</td>\n",
       "      <td>male</td>\n",
       "      <td>SAME124182</td>\n",
       "      <td>GIH</td>\n",
       "      <td>Gujarati</td>\n",
       "      <td>SAS</td>\n",
       "      <td>South Asian Ancestry</td>\n",
       "      <td>GIH</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2706</th>\n",
       "      <td>NA21117</td>\n",
       "      <td>male</td>\n",
       "      <td>SAME124185</td>\n",
       "      <td>GIH</td>\n",
       "      <td>Gujarati</td>\n",
       "      <td>SAS</td>\n",
       "      <td>South Asian Ancestry</td>\n",
       "      <td>GIH</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2707</th>\n",
       "      <td>NA21124</td>\n",
       "      <td>male</td>\n",
       "      <td>SAME124024</td>\n",
       "      <td>GIH</td>\n",
       "      <td>Gujarati</td>\n",
       "      <td>SAS</td>\n",
       "      <td>South Asian Ancestry</td>\n",
       "      <td>GIH</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2708</th>\n",
       "      <td>NA21129</td>\n",
       "      <td>male</td>\n",
       "      <td>SAME124023</td>\n",
       "      <td>GIH</td>\n",
       "      <td>Gujarati</td>\n",
       "      <td>SAS</td>\n",
       "      <td>South Asian Ancestry</td>\n",
       "      <td>GIH</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Sample name     Sex Biosample ID Population code Population name  \\\n",
       "2694     NA20897    male   SAME123664             GIH        Gujarati   \n",
       "2695     NA20900  female   SAME122920             GIH        Gujarati   \n",
       "2696     NA20905    male   SAME122915             GIH        Gujarati   \n",
       "2697     NA21092    male   SAME123322             GIH        Gujarati   \n",
       "2698     NA21097  female   SAME123319             GIH        Gujarati   \n",
       "2699     NA21100    male   SAME124440             GIH        Gujarati   \n",
       "2700     NA21105    male   SAME124437             GIH        Gujarati   \n",
       "2701     NA20873    male   SAME122841             GIH        Gujarati   \n",
       "2702     NA20878  female   SAME123595             GIH        Gujarati   \n",
       "2703     NA20885    male   SAME124525             GIH        Gujarati   \n",
       "2704     NA21143  female   SAME123359             GIH        Gujarati   \n",
       "2705     NA21112    male   SAME124182             GIH        Gujarati   \n",
       "2706     NA21117    male   SAME124185             GIH        Gujarati   \n",
       "2707     NA21124    male   SAME124024             GIH        Gujarati   \n",
       "2708     NA21129    male   SAME124023             GIH        Gujarati   \n",
       "\n",
       "     Superpopulation code  Superpopulation name Population elastic ID  \\\n",
       "2694                  SAS  South Asian Ancestry                   GIH   \n",
       "2695                  SAS  South Asian Ancestry                   GIH   \n",
       "2696                  SAS  South Asian Ancestry                   GIH   \n",
       "2697                  SAS  South Asian Ancestry                   GIH   \n",
       "2698                  SAS  South Asian Ancestry                   GIH   \n",
       "2699                  SAS  South Asian Ancestry                   GIH   \n",
       "2700                  SAS  South Asian Ancestry                   GIH   \n",
       "2701                  SAS  South Asian Ancestry                   GIH   \n",
       "2702                  SAS  South Asian Ancestry                   GIH   \n",
       "2703                  SAS  South Asian Ancestry                   GIH   \n",
       "2704                  SAS  South Asian Ancestry                   GIH   \n",
       "2705                  SAS  South Asian Ancestry                   GIH   \n",
       "2706                  SAS  South Asian Ancestry                   GIH   \n",
       "2707                  SAS  South Asian Ancestry                   GIH   \n",
       "2708                  SAS  South Asian Ancestry                   GIH   \n",
       "\n",
       "                                       Data collections  \n",
       "2694  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "2695  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "2696  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "2697  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "2698  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "2699  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "2700  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "2701  1000 Genomes on GRCh38,1000 Genomes phase 3 re...  \n",
       "2702  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "2703  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "2704  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "2705  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "2706  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "2707  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "2708  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  "
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_df.tail(15)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Summary Statistics\n",
    "The .describe() method provides summary statistics of the DataFrame, useful for getting a sense of the distribution of each attribute."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample name</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Biosample ID</th>\n",
       "      <th>Population code</th>\n",
       "      <th>Population name</th>\n",
       "      <th>Superpopulation code</th>\n",
       "      <th>Superpopulation name</th>\n",
       "      <th>Population elastic ID</th>\n",
       "      <th>Data collections</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>2709</td>\n",
       "      <td>2709</td>\n",
       "      <td>2709</td>\n",
       "      <td>2709</td>\n",
       "      <td>2709</td>\n",
       "      <td>2709</td>\n",
       "      <td>2709</td>\n",
       "      <td>2709</td>\n",
       "      <td>2709</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique</th>\n",
       "      <td>2709</td>\n",
       "      <td>2</td>\n",
       "      <td>2709</td>\n",
       "      <td>27</td>\n",
       "      <td>38</td>\n",
       "      <td>6</td>\n",
       "      <td>10</td>\n",
       "      <td>38</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>top</th>\n",
       "      <td>HG00271</td>\n",
       "      <td>female</td>\n",
       "      <td>SAME123417</td>\n",
       "      <td>GWD</td>\n",
       "      <td>Gambian Mandinka</td>\n",
       "      <td>AFR</td>\n",
       "      <td>African Ancestry</td>\n",
       "      <td>GWD</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>freq</th>\n",
       "      <td>1</td>\n",
       "      <td>1379</td>\n",
       "      <td>1</td>\n",
       "      <td>120</td>\n",
       "      <td>118</td>\n",
       "      <td>719</td>\n",
       "      <td>711</td>\n",
       "      <td>118</td>\n",
       "      <td>1251</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Sample name     Sex Biosample ID Population code   Population name  \\\n",
       "count         2709    2709         2709            2709              2709   \n",
       "unique        2709       2         2709              27                38   \n",
       "top        HG00271  female   SAME123417             GWD  Gambian Mandinka   \n",
       "freq             1    1379            1             120               118   \n",
       "\n",
       "       Superpopulation code Superpopulation name Population elastic ID  \\\n",
       "count                  2709                 2709                  2709   \n",
       "unique                    6                   10                    38   \n",
       "top                     AFR     African Ancestry                   GWD   \n",
       "freq                    719                  711                   118   \n",
       "\n",
       "                                         Data collections  \n",
       "count                                                2709  \n",
       "unique                                                 33  \n",
       "top     1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "freq                                                 1251  "
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get summary statistics for the sample DataFrame\n",
    "sample_df.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Count Values\n",
    "To count the number of occurrences of each unique value in a column, you can use .value_counts()."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Population name\n",
       "Gambian Mandinka            118\n",
       "Southern Han Chinese        114\n",
       "Telugu                      112\n",
       "Toscani                     112\n",
       "Yoruba                      111\n",
       "Luhya                       111\n",
       "Tamil                       111\n",
       "Dai Chinese                 109\n",
       "Gujarati                    109\n",
       "Esan                        109\n",
       "Han Chinese                 108\n",
       "Puerto Rican                107\n",
       "Iberian                     105\n",
       "Colombian                   105\n",
       "Japanese                    104\n",
       "Punjabi                     104\n",
       "British                     104\n",
       "CEPH                        103\n",
       "Finnish                     102\n",
       "Bengali                     102\n",
       "Kinh Vietnamese             101\n",
       "African Caribbean            98\n",
       "Mende                        96\n",
       "Peruvian                     91\n",
       "Mexican Ancestry             70\n",
       "African Ancestry SW          68\n",
       "Punjabi,Punjabi               4\n",
       "Finnish,Finnish               3\n",
       "British,English               2\n",
       "Bengali,Bengali               2\n",
       "Kinh,Kinh Vietnamese          2\n",
       "Mende,Mende                   2\n",
       "Luhya,Luhya                   2\n",
       "Esan,Esan                     2\n",
       "Gambian Mandinka,Gambian      2\n",
       "Iberian,Spanish               2\n",
       "Japanese,Japanese             1\n",
       "Iberian,Mende                 1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Count the number of individuals per population\n",
    "sample_df['Population name'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filtering Data\n",
    "You can filter rows based on certain conditions. For example, let's filter the sample DataFrame to only include individuals from a specific population."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample name</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Biosample ID</th>\n",
       "      <th>Population code</th>\n",
       "      <th>Population name</th>\n",
       "      <th>Superpopulation code</th>\n",
       "      <th>Superpopulation name</th>\n",
       "      <th>Population elastic ID</th>\n",
       "      <th>Data collections</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>711</th>\n",
       "      <td>NA18489</td>\n",
       "      <td>female</td>\n",
       "      <td>SAME124102</td>\n",
       "      <td>YRI</td>\n",
       "      <td>Yoruba</td>\n",
       "      <td>AFR</td>\n",
       "      <td>African Ancestry</td>\n",
       "      <td>YRI</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>713</th>\n",
       "      <td>NA18504</td>\n",
       "      <td>male</td>\n",
       "      <td>SAME125061</td>\n",
       "      <td>YRI</td>\n",
       "      <td>Yoruba</td>\n",
       "      <td>AFR</td>\n",
       "      <td>African Ancestry</td>\n",
       "      <td>YRI</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>716</th>\n",
       "      <td>NA18511</td>\n",
       "      <td>female</td>\n",
       "      <td>SAME124885</td>\n",
       "      <td>YRI</td>\n",
       "      <td>Yoruba</td>\n",
       "      <td>AFR</td>\n",
       "      <td>African Ancestry</td>\n",
       "      <td>YRI</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>717</th>\n",
       "      <td>NA18516</td>\n",
       "      <td>male</td>\n",
       "      <td>SAME124884</td>\n",
       "      <td>YRI</td>\n",
       "      <td>Yoruba</td>\n",
       "      <td>AFR</td>\n",
       "      <td>African Ancestry</td>\n",
       "      <td>YRI</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>740</th>\n",
       "      <td>NA18523</td>\n",
       "      <td>female</td>\n",
       "      <td>SAME124670</td>\n",
       "      <td>YRI</td>\n",
       "      <td>Yoruba</td>\n",
       "      <td>AFR</td>\n",
       "      <td>African Ancestry</td>\n",
       "      <td>YRI</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Sample name     Sex Biosample ID Population code Population name  \\\n",
       "711     NA18489  female   SAME124102             YRI          Yoruba   \n",
       "713     NA18504    male   SAME125061             YRI          Yoruba   \n",
       "716     NA18511  female   SAME124885             YRI          Yoruba   \n",
       "717     NA18516    male   SAME124884             YRI          Yoruba   \n",
       "740     NA18523  female   SAME124670             YRI          Yoruba   \n",
       "\n",
       "    Superpopulation code Superpopulation name Population elastic ID  \\\n",
       "711                  AFR     African Ancestry                   YRI   \n",
       "713                  AFR     African Ancestry                   YRI   \n",
       "716                  AFR     African Ancestry                   YRI   \n",
       "717                  AFR     African Ancestry                   YRI   \n",
       "740                  AFR     African Ancestry                   YRI   \n",
       "\n",
       "                                      Data collections  \n",
       "711  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "713  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "716  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "717  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "740  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  "
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter to include only individuals from the 'YRI' population\n",
    "yri_population = sample_df[sample_df['Population code'] == 'YRI']\n",
    "yri_population.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pandas.core.frame.DataFrame"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(yri_population)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Index: 111 entries, 711 to 2651\n",
      "Data columns (total 9 columns):\n",
      " #   Column                 Non-Null Count  Dtype \n",
      "---  ------                 --------------  ----- \n",
      " 0   Sample name            111 non-null    object\n",
      " 1   Sex                    111 non-null    object\n",
      " 2   Biosample ID           111 non-null    object\n",
      " 3   Population code        111 non-null    object\n",
      " 4   Population name        111 non-null    object\n",
      " 5   Superpopulation code   111 non-null    object\n",
      " 6   Superpopulation name   111 non-null    object\n",
      " 7   Population elastic ID  111 non-null    object\n",
      " 8   Data collections       111 non-null    object\n",
      "dtypes: object(9)\n",
      "memory usage: 8.7+ KB\n"
     ]
    }
   ],
   "source": [
    "yri_population.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample name</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Biosample ID</th>\n",
       "      <th>Population code</th>\n",
       "      <th>Population name</th>\n",
       "      <th>Superpopulation code</th>\n",
       "      <th>Superpopulation name</th>\n",
       "      <th>Population elastic ID</th>\n",
       "      <th>Data collections</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>111</td>\n",
       "      <td>111</td>\n",
       "      <td>111</td>\n",
       "      <td>111</td>\n",
       "      <td>111</td>\n",
       "      <td>111</td>\n",
       "      <td>111</td>\n",
       "      <td>111</td>\n",
       "      <td>111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique</th>\n",
       "      <td>111</td>\n",
       "      <td>2</td>\n",
       "      <td>111</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>top</th>\n",
       "      <td>NA18489</td>\n",
       "      <td>female</td>\n",
       "      <td>SAME124102</td>\n",
       "      <td>YRI</td>\n",
       "      <td>Yoruba</td>\n",
       "      <td>AFR</td>\n",
       "      <td>African Ancestry</td>\n",
       "      <td>YRI</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>freq</th>\n",
       "      <td>1</td>\n",
       "      <td>57</td>\n",
       "      <td>1</td>\n",
       "      <td>111</td>\n",
       "      <td>111</td>\n",
       "      <td>111</td>\n",
       "      <td>111</td>\n",
       "      <td>111</td>\n",
       "      <td>89</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Sample name     Sex Biosample ID Population code Population name  \\\n",
       "count          111     111          111             111             111   \n",
       "unique         111       2          111               1               1   \n",
       "top        NA18489  female   SAME124102             YRI          Yoruba   \n",
       "freq             1      57            1             111             111   \n",
       "\n",
       "       Superpopulation code Superpopulation name Population elastic ID  \\\n",
       "count                   111                  111                   111   \n",
       "unique                    1                    1                     1   \n",
       "top                     AFR     African Ancestry                   YRI   \n",
       "freq                    111                  111                   111   \n",
       "\n",
       "                                         Data collections  \n",
       "count                                                 111  \n",
       "unique                                                  4  \n",
       "top     1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "freq                                                   89  "
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "yri_population.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Index: 57 entries, 711 to 2651\n",
      "Data columns (total 9 columns):\n",
      " #   Column                 Non-Null Count  Dtype \n",
      "---  ------                 --------------  ----- \n",
      " 0   Sample name            57 non-null     object\n",
      " 1   Sex                    57 non-null     object\n",
      " 2   Biosample ID           57 non-null     object\n",
      " 3   Population code        57 non-null     object\n",
      " 4   Population name        57 non-null     object\n",
      " 5   Superpopulation code   57 non-null     object\n",
      " 6   Superpopulation name   57 non-null     object\n",
      " 7   Population elastic ID  57 non-null     object\n",
      " 8   Data collections       57 non-null     object\n",
      "dtypes: object(9)\n",
      "memory usage: 4.5+ KB\n"
     ]
    }
   ],
   "source": [
    "# Search for individuals with specific attributes\n",
    "specific_entries = sample_df[(sample_df['Population code'] == 'YRI') & (sample_df['Sex'] == 'female')]\n",
    "specific_entries.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample name</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Biosample ID</th>\n",
       "      <th>Population code</th>\n",
       "      <th>Population name</th>\n",
       "      <th>Superpopulation code</th>\n",
       "      <th>Superpopulation name</th>\n",
       "      <th>Population elastic ID</th>\n",
       "      <th>Data collections</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>711</th>\n",
       "      <td>NA18489</td>\n",
       "      <td>female</td>\n",
       "      <td>SAME124102</td>\n",
       "      <td>YRI</td>\n",
       "      <td>Yoruba</td>\n",
       "      <td>AFR</td>\n",
       "      <td>African Ancestry</td>\n",
       "      <td>YRI</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>716</th>\n",
       "      <td>NA18511</td>\n",
       "      <td>female</td>\n",
       "      <td>SAME124885</td>\n",
       "      <td>YRI</td>\n",
       "      <td>Yoruba</td>\n",
       "      <td>AFR</td>\n",
       "      <td>African Ancestry</td>\n",
       "      <td>YRI</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>740</th>\n",
       "      <td>NA18523</td>\n",
       "      <td>female</td>\n",
       "      <td>SAME124670</td>\n",
       "      <td>YRI</td>\n",
       "      <td>Yoruba</td>\n",
       "      <td>AFR</td>\n",
       "      <td>African Ancestry</td>\n",
       "      <td>YRI</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>752</th>\n",
       "      <td>NA18508</td>\n",
       "      <td>female</td>\n",
       "      <td>SAME124318</td>\n",
       "      <td>YRI</td>\n",
       "      <td>Yoruba</td>\n",
       "      <td>AFR</td>\n",
       "      <td>African Ancestry</td>\n",
       "      <td>YRI</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>770</th>\n",
       "      <td>NA18876</td>\n",
       "      <td>female</td>\n",
       "      <td>SAME125103</td>\n",
       "      <td>YRI</td>\n",
       "      <td>Yoruba</td>\n",
       "      <td>AFR</td>\n",
       "      <td>African Ancestry</td>\n",
       "      <td>YRI</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Sample name     Sex Biosample ID Population code Population name  \\\n",
       "711     NA18489  female   SAME124102             YRI          Yoruba   \n",
       "716     NA18511  female   SAME124885             YRI          Yoruba   \n",
       "740     NA18523  female   SAME124670             YRI          Yoruba   \n",
       "752     NA18508  female   SAME124318             YRI          Yoruba   \n",
       "770     NA18876  female   SAME125103             YRI          Yoruba   \n",
       "\n",
       "    Superpopulation code Superpopulation name Population elastic ID  \\\n",
       "711                  AFR     African Ancestry                   YRI   \n",
       "716                  AFR     African Ancestry                   YRI   \n",
       "740                  AFR     African Ancestry                   YRI   \n",
       "752                  AFR     African Ancestry                   YRI   \n",
       "770                  AFR     African Ancestry                   YRI   \n",
       "\n",
       "                                      Data collections  \n",
       "711  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "716  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "740  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "752  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "770  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  "
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "specific_entries.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample name</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Biosample ID</th>\n",
       "      <th>Population code</th>\n",
       "      <th>Population name</th>\n",
       "      <th>Superpopulation code</th>\n",
       "      <th>Superpopulation name</th>\n",
       "      <th>Population elastic ID</th>\n",
       "      <th>Data collections</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique</th>\n",
       "      <td>57</td>\n",
       "      <td>1</td>\n",
       "      <td>57</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>top</th>\n",
       "      <td>NA18489</td>\n",
       "      <td>female</td>\n",
       "      <td>SAME124102</td>\n",
       "      <td>YRI</td>\n",
       "      <td>Yoruba</td>\n",
       "      <td>AFR</td>\n",
       "      <td>African Ancestry</td>\n",
       "      <td>YRI</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>freq</th>\n",
       "      <td>1</td>\n",
       "      <td>57</td>\n",
       "      <td>1</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Sample name     Sex Biosample ID Population code Population name  \\\n",
       "count           57      57           57              57              57   \n",
       "unique          57       1           57               1               1   \n",
       "top        NA18489  female   SAME124102             YRI          Yoruba   \n",
       "freq             1      57            1              57              57   \n",
       "\n",
       "       Superpopulation code Superpopulation name Population elastic ID  \\\n",
       "count                    57                   57                    57   \n",
       "unique                    1                    1                     1   \n",
       "top                     AFR     African Ancestry                   YRI   \n",
       "freq                     57                   57                    57   \n",
       "\n",
       "                                         Data collections  \n",
       "count                                                  57  \n",
       "unique                                                  4  \n",
       "top     1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "freq                                                   49  "
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "specific_entries.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Multiple Conditions\n",
    "You can include multiple conditions in your query. For example, let's find all females in either the 'YRI' or 'CEU' populations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample name</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Biosample ID</th>\n",
       "      <th>Population code</th>\n",
       "      <th>Population name</th>\n",
       "      <th>Superpopulation code</th>\n",
       "      <th>Superpopulation name</th>\n",
       "      <th>Population elastic ID</th>\n",
       "      <th>Data collections</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>110</td>\n",
       "      <td>110</td>\n",
       "      <td>110</td>\n",
       "      <td>110</td>\n",
       "      <td>110</td>\n",
       "      <td>110</td>\n",
       "      <td>110</td>\n",
       "      <td>110</td>\n",
       "      <td>110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique</th>\n",
       "      <td>110</td>\n",
       "      <td>1</td>\n",
       "      <td>110</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>top</th>\n",
       "      <td>NA12828</td>\n",
       "      <td>female</td>\n",
       "      <td>SAME125076</td>\n",
       "      <td>YRI</td>\n",
       "      <td>Yoruba</td>\n",
       "      <td>AFR</td>\n",
       "      <td>African Ancestry</td>\n",
       "      <td>YRI</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>freq</th>\n",
       "      <td>1</td>\n",
       "      <td>110</td>\n",
       "      <td>1</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>95</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Sample name     Sex Biosample ID Population code Population name  \\\n",
       "count          110     110          110             110             110   \n",
       "unique         110       1          110               2               2   \n",
       "top        NA12828  female   SAME125076             YRI          Yoruba   \n",
       "freq             1     110            1              57              57   \n",
       "\n",
       "       Superpopulation code Superpopulation name Population elastic ID  \\\n",
       "count                   110                  110                   110   \n",
       "unique                    2                    2                     2   \n",
       "top                     AFR     African Ancestry                   YRI   \n",
       "freq                     57                   57                    57   \n",
       "\n",
       "                                         Data collections  \n",
       "count                                                 110  \n",
       "unique                                                  7  \n",
       "top     1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "freq                                                   95  "
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Search for females in either 'YRI' or 'CEU' populations\n",
    "multiple_conditions = sample_df[(sample_df['Population code'].isin(['YRI', 'CEU'])) & (sample_df['Sex'] == 'female')]\n",
    "multiple_conditions.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample name</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Biosample ID</th>\n",
       "      <th>Population code</th>\n",
       "      <th>Population name</th>\n",
       "      <th>Superpopulation code</th>\n",
       "      <th>Superpopulation name</th>\n",
       "      <th>Population elastic ID</th>\n",
       "      <th>Data collections</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique</th>\n",
       "      <td>57</td>\n",
       "      <td>1</td>\n",
       "      <td>57</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>top</th>\n",
       "      <td>NA18489</td>\n",
       "      <td>female</td>\n",
       "      <td>SAME124102</td>\n",
       "      <td>YRI</td>\n",
       "      <td>Yoruba</td>\n",
       "      <td>AFR</td>\n",
       "      <td>African Ancestry</td>\n",
       "      <td>YRI</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>freq</th>\n",
       "      <td>1</td>\n",
       "      <td>57</td>\n",
       "      <td>1</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Sample name     Sex Biosample ID Population code Population name  \\\n",
       "count           57      57           57              57              57   \n",
       "unique          57       1           57               1               1   \n",
       "top        NA18489  female   SAME124102             YRI          Yoruba   \n",
       "freq             1      57            1              57              57   \n",
       "\n",
       "       Superpopulation code Superpopulation name Population elastic ID  \\\n",
       "count                    57                   57                    57   \n",
       "unique                    1                    1                     1   \n",
       "top                     AFR     African Ancestry                   YRI   \n",
       "freq                     57                   57                    57   \n",
       "\n",
       "                                         Data collections  \n",
       "count                                                  57  \n",
       "unique                                                  4  \n",
       "top     1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "freq                                                   49  "
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "target_population = 'YRI'\n",
    "target_sex = 'female'\n",
    "\n",
    "# Search using variables\n",
    "variable_filter = sample_df[(sample_df['Population code'] == target_population) & (sample_df['Sex'] == target_sex)]\n",
    "variable_filter.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Searching for Entries in a List\n",
    "If you have a list of values to search for, you can use the isin() method within .query()."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample name</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Biosample ID</th>\n",
       "      <th>Population code</th>\n",
       "      <th>Population name</th>\n",
       "      <th>Superpopulation code</th>\n",
       "      <th>Superpopulation name</th>\n",
       "      <th>Population elastic ID</th>\n",
       "      <th>Data collections</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>214</td>\n",
       "      <td>214</td>\n",
       "      <td>214</td>\n",
       "      <td>214</td>\n",
       "      <td>214</td>\n",
       "      <td>214</td>\n",
       "      <td>214</td>\n",
       "      <td>214</td>\n",
       "      <td>214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique</th>\n",
       "      <td>214</td>\n",
       "      <td>2</td>\n",
       "      <td>214</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>top</th>\n",
       "      <td>NA12828</td>\n",
       "      <td>female</td>\n",
       "      <td>SAME125076</td>\n",
       "      <td>YRI</td>\n",
       "      <td>Yoruba</td>\n",
       "      <td>AFR</td>\n",
       "      <td>African Ancestry</td>\n",
       "      <td>YRI</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>freq</th>\n",
       "      <td>1</td>\n",
       "      <td>110</td>\n",
       "      <td>1</td>\n",
       "      <td>111</td>\n",
       "      <td>111</td>\n",
       "      <td>111</td>\n",
       "      <td>111</td>\n",
       "      <td>111</td>\n",
       "      <td>179</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Sample name     Sex Biosample ID Population code Population name  \\\n",
       "count          214     214          214             214             214   \n",
       "unique         214       2          214               2               2   \n",
       "top        NA12828  female   SAME125076             YRI          Yoruba   \n",
       "freq             1     110            1             111             111   \n",
       "\n",
       "       Superpopulation code Superpopulation name Population elastic ID  \\\n",
       "count                   214                  214                   214   \n",
       "unique                    2                    2                     2   \n",
       "top                     AFR     African Ancestry                   YRI   \n",
       "freq                    111                  111                   111   \n",
       "\n",
       "                                         Data collections  \n",
       "count                                                 214  \n",
       "unique                                                  7  \n",
       "top     1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "freq                                                  179  "
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# List of target populations\n",
    "target_populations = ['YRI', 'CEU']\n",
    "\n",
    "# Search for individuals in target populations\n",
    "list_filter = sample_df[sample_df['Population code'].isin(target_populations)]\n",
    "list_filter.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Using String Methods\n",
    "You can also use string methods to search for specific patterns in string columns."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "899\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample name</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Biosample ID</th>\n",
       "      <th>Population code</th>\n",
       "      <th>Population name</th>\n",
       "      <th>Superpopulation code</th>\n",
       "      <th>Superpopulation name</th>\n",
       "      <th>Population elastic ID</th>\n",
       "      <th>Data collections</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>698</th>\n",
       "      <td>NA12828</td>\n",
       "      <td>female</td>\n",
       "      <td>SAME125076</td>\n",
       "      <td>CEU</td>\n",
       "      <td>CEPH</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European Ancestry</td>\n",
       "      <td>CEU</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>700</th>\n",
       "      <td>NA12830</td>\n",
       "      <td>female</td>\n",
       "      <td>SAME125289</td>\n",
       "      <td>CEU</td>\n",
       "      <td>CEPH</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European Ancestry</td>\n",
       "      <td>CEU</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>702</th>\n",
       "      <td>NA12842</td>\n",
       "      <td>male</td>\n",
       "      <td>SAME123542</td>\n",
       "      <td>CEU</td>\n",
       "      <td>CEPH</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European Ancestry</td>\n",
       "      <td>CEU</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>704</th>\n",
       "      <td>NA12873</td>\n",
       "      <td>female</td>\n",
       "      <td>SAME123388</td>\n",
       "      <td>CEU</td>\n",
       "      <td>CEPH</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European Ancestry</td>\n",
       "      <td>CEU</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>706</th>\n",
       "      <td>NA12878</td>\n",
       "      <td>female</td>\n",
       "      <td>SAME123392</td>\n",
       "      <td>CEU</td>\n",
       "      <td>CEPH</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European Ancestry</td>\n",
       "      <td>CEU</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Sample name     Sex Biosample ID Population code Population name  \\\n",
       "698     NA12828  female   SAME125076             CEU            CEPH   \n",
       "700     NA12830  female   SAME125289             CEU            CEPH   \n",
       "702     NA12842    male   SAME123542             CEU            CEPH   \n",
       "704     NA12873  female   SAME123388             CEU            CEPH   \n",
       "706     NA12878  female   SAME123392             CEU            CEPH   \n",
       "\n",
       "    Superpopulation code Superpopulation name Population elastic ID  \\\n",
       "698                  EUR    European Ancestry                   CEU   \n",
       "700                  EUR    European Ancestry                   CEU   \n",
       "702                  EUR    European Ancestry                   CEU   \n",
       "704                  EUR    European Ancestry                   CEU   \n",
       "706                  EUR    European Ancestry                   CEU   \n",
       "\n",
       "                                      Data collections  \n",
       "698  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "700  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "702  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "704  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "706  1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  "
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Search for individuals whose sample IDs start with 'NA'\n",
    "string_filter = sample_df[sample_df['Sample name'].str.startswith('NA')]\n",
    "print(len(string_filter))\n",
    "string_filter.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exploratory Data Analysis\n",
    "\n",
    "Before we move on, let's specifically look at:\n",
    "\n",
    "1. The total number of samples in the dataset.\n",
    "2. The distribution of samples by sex.\n",
    "3. The distribution of samples by population.\n",
    "4. The distribution of samples by superpopulation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of samples: 2709\n",
      "\n",
      "Distribution of samples by Sex:\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Sex\n",
       "female    1379\n",
       "male      1330\n",
       "Name: count, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Distribution of samples by Superpopulation:\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Superpopulation name\n",
       "African Ancestry                          711\n",
       "South Asian Ancestry                      538\n",
       "East Asian Ancestry                       536\n",
       "European Ancestry                         526\n",
       "American Ancestry                         373\n",
       "African Ancestry,Africa (SGDP)              8\n",
       "European Ancestry,West Eurasia (SGDP)       7\n",
       "South Asia (SGDP),South Asian Ancestry      6\n",
       "East Asia (SGDP),East Asian Ancestry        3\n",
       "European Ancestry,African Ancestry          1\n",
       "Name: count, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Check if the sample DataFrame is loaded\n",
    "if 'sample_df' in locals():\n",
    "\n",
    "    # Total number of samples\n",
    "    total_samples = len(sample_df)\n",
    "    print(f\"Total number of samples: {total_samples}\")\n",
    "\n",
    "    # Distribution by Sex\n",
    "    print(\"\\nDistribution of samples by Sex:\")\n",
    "    display(sample_df['Sex'].value_counts())\n",
    "\n",
    "    # Distribution by Superpopulation\n",
    "    print(\"\\nDistribution of samples by Superpopulation:\")\n",
    "    display(sample_df['Superpopulation name'].value_counts())\n",
    "\n",
    "else:\n",
    "    print(\"The sample DataFrame is not loaded. Please make sure to load the sample file.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Downloading 1000 Genomes Data\n",
    "\n",
    "Up to this point, we have explored the sample and population files to understand the structure of our data. However, we have not yet downloaded the actual genetic data from the 1000 Genomes Project. \n",
    "\n",
    "In the next step, we will download the VCF (Variant Call Format) files for each chromosome. These files contain the genetic variants for our samples and are essential for our subsequent analyses.\n",
    "\n",
    "Note: The script will also check if the directory for storing the 1000 Genomes reference panel exists. If not, it will create one for you.\n",
    "\n",
    "Let's proceed to download the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "These files are very large so expect to let this run about a day.\n",
      "\n",
      "Downloading chromosome 1...\n",
      "Downloading chromosome 2...\n",
      "Downloading chromosome 3...\n",
      "Downloading chromosome 4...\n",
      "Downloading chromosome 5...\n",
      "Downloading chromosome 6...\n",
      "Downloading chromosome 7...\n",
      "Downloading chromosome 8...\n",
      "Downloading chromosome 9...\n",
      "Downloading chromosome 10...\n",
      "Downloading chromosome 11...\n",
      "Downloading chromosome 12...\n",
      "Downloading chromosome 13...\n",
      "Downloading chromosome 14...\n",
      "Downloading chromosome 15...\n",
      "Downloading chromosome 16...\n",
      "Downloading chromosome 17...\n",
      "Downloading chromosome 18...\n",
      "Downloading chromosome 19...\n",
      "Downloading chromosome 20...\n",
      "Downloading chromosome 21...\n",
      "Downloading chromosome 22...\n"
     ]
    }
   ],
   "source": [
    "%%bash -s \"$primary_directory\" \"$references_directory\" \"$results_directory\"\n",
    "\n",
    "# Receive directory variables from Python\n",
    "primary_dir=$1\n",
    "ref_dir=$2\n",
    "results_dir=$3\n",
    "\n",
    "# Define the directory for the 1000 Genomes reference panel\n",
    "onekg_reference_panel_dir=\"${ref_dir}/onekg_reference_panel\"\n",
    "\n",
    "# Check if the onekg_reference_panel directory exists; if not, create it\n",
    "if [ ! -d \"${onekg_reference_panel_dir}\" ]; then\n",
    "    echo \"Creating onekg_reference_panel directory...\"\n",
    "    mkdir -p \"${onekg_reference_panel_dir}\"\n",
    "fi\n",
    "\n",
    "echo \"These files are very large so expect to let this run for several hours.\"\n",
    "echo\n",
    "\n",
    "# Loop through each chromosome to download data\n",
    "for chromosome in {1..22}\n",
    "do\n",
    "    # Download the VCF file for each chromosome from the 1000 Genomes FTP site\n",
    "    echo \"Downloading chromosome ${chromosome}...\"\n",
    "    wget --quiet ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/1000_genomes_project/release/20190312_biallelic_SNV_and_INDEL/ALL.chr${chromosome}.shapeit2_integrated_snvindels_v2a_27022019.GRCh38.phased.vcf.gz \\\n",
    "        -P ${onekg_reference_panel_dir}\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Subsetting to 500 Individuals\n",
    "\n",
    "### Why Is It Necessary?\n",
    "\n",
    "As we dive deeper into genetic data analysis, it's important to manage computational resources effectively. Whole-genome VCF files from projects like the 1000 Genomes can be extremely large, often containing data for thousands of individuals across millions of genetic variants. Loading such extensive data into memory can be computationally intensive and may even cause the kernel to crash, as we've experienced.\n",
    "\n",
    "### Advantages of Subsetting\n",
    "\n",
    "1. **Reduced Computational Load**: By focusing on a subset of 500 individuals, we significantly reduce the computational resources needed for the analysis.\n",
    "\n",
    "2. **Faster Execution**: Smaller datasets mean that code will execute more quickly, allowing us to focus on the analysis rather than waiting for code to run.\n",
    "\n",
    "3. **Feasibility**: Not all personal computers will have the resources to handle large genomic datasets. Subsetting makes the tutorial more accessible to individuals with less powerful machines.\n",
    "\n",
    "4. **Focused Analysis**: With fewer individuals, it's easier to explore the data in depth, which is particularly useful for educational purposes.\n",
    "\n",
    "### Random Selection\n",
    "\n",
    "The 500 individuals are randomly selected to ensure that the subset is representative of the larger dataset. This allows us to make more general conclusions from our analyses.\n",
    "\n",
    "In the next cell, we will use `bcftools` to subset our VCF file to include only these 500 randomly selected individuals. This will make subsequent analyses more manageable and less resource-intensive."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/bin/bcftools\n",
      "bcftools is already installed.\n"
     ]
    }
   ],
   "source": [
    "# Ubuntu\n",
    "import os\n",
    "\n",
    "# Check if bcftools is installed\n",
    "bcftools_installed = os.system(\"which bcftools\")\n",
    "\n",
    "# Install bcftools if not installed\n",
    "if bcftools_installed != 0:\n",
    "    print(\"bcftools is not installed. Get in instructions on how to install.\")\n",
    "    # os.system(\"sudo apt-get update\")\n",
    "    # os.system(\"sudo apt-get install -y bcftools\")\n",
    "else:\n",
    "    print(\"bcftools is already installed.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$references_directory\"\n",
    "\n",
    "# Receive directory variables from Python\n",
    "ref_dir=$1\n",
    "\n",
    "# Define the directory for the 1000 Genomes reference panel\n",
    "onekg_reference_panel_dir=\"${ref_dir}/onekg_reference_panel\"\n",
    "\n",
    "# Extract sample names from the VCF header\n",
    "bcftools query -l ${onekg_reference_panel_dir}/ALL.chr20.shapeit2_integrated_snvindels_v2a_27022019.GRCh38.phased.vcf.gz > ${onekg_reference_panel_dir}/all_sample_ids.txt\n",
    "\n",
    "# Randomly select 500 sample names\n",
    "shuf -n 500 ${onekg_reference_panel_dir}/all_sample_ids.txt > ${onekg_reference_panel_dir}/random_500_sample_ids.txt\n",
    "\n",
    "# Subset the VCF file based on the random 500 sample IDs\n",
    "bcftools view -S ${onekg_reference_panel_dir}/random_500_sample_ids.txt -o ${onekg_reference_panel_dir}/subset_500_chr20.vcf.gz -Oz ${onekg_reference_panel_dir}/ALL.chr20.shapeit2_integrated_snvindels_v2a_27022019.GRCh38.phased.vcf.gz\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exploring the VCF Files\n",
    "\n",
    "Now that we have downloaded the VCF files and selected a subset of 500 indiviuals, it's time to explore the VCF files. VCF (Variant Call Format) files contain information about genetic variants found in the samples.\n",
    "\n",
    "In this section, we will:\n",
    "1. Load a VCF file for a specific chromosome.\n",
    "2. Explore the structure of the VCF file.\n",
    "3. Examine some basic statistics.\n",
    "\n",
    "Let's get started!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Installing Required Packages\n",
    "\n",
    "In data analysis and scientific computing, we often rely on specialized packages to perform specific tasks. These packages are collections of functions and methods that allow us to perform operations without having to write code from scratch.\n",
    "\n",
    "Before we can use these packages, we need to install them. This is usually a one-time operation. Below, we will install the `scikit-allel` package, which provides tools for bioinformatics and genomics, particularly in the domain of high-throughput sequencing.\n",
    "\n",
    "Let's proceed to install the package."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Loading a VCF File\n",
    "\n",
    "We will start by loading a VCF file for a specific chromosome. For demonstration purposes, let's focus on chromosome 1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Keys in the VCF file: ['samples', 'calldata/GT', 'variants/ALT', 'variants/CHROM', 'variants/FILTER_PASS', 'variants/ID', 'variants/POS', 'variants/QUAL', 'variants/REF']\n",
      "Notice the output are the keys in the VCF file. We will use the second key, 'calldata/GT', 'variants/ALT', and 'variants/REF'\n"
     ]
    }
   ],
   "source": [
    "import allel  # Importing the scikit-allel package\n",
    "\n",
    "# Define the path to the subsetted VCF file\n",
    "vcf_path_chromosome_20_subset = os.path.join(references_directory, \"onekg_reference_panel\", \"subset_500_chr20.vcf.gz\")\n",
    "\n",
    "# Check if the subsetted VCF file exists\n",
    "if os.path.exists(vcf_path_chromosome_20_subset):\n",
    "    # Load the subsetted VCF file\n",
    "    callset = allel.read_vcf(vcf_path_chromosome_20_subset)\n",
    "    \n",
    "    # Display the keys to understand the structure\n",
    "    print(\"Keys in the VCF file:\", list(callset.keys()))\n",
    "else:\n",
    "    print(f\"The subsetted VCF file does not exist at {vcf_path_chromosome_20_subset}. Please make sure the file is in the correct location.\")\n",
    "\n",
    "# NOTE: add instructions for if it crash here.\n",
    "\n",
    "print(\"Notice the output are the keys in the VCF file. We will use the second key, 'calldata/GT', 'variants/ALT', and 'variants/REF'\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Understanding Allele Frequencies and SNPs\n",
    "\n",
    "Before we proceed to subset our data, it's essential to understand some key genetic concepts: Allele Frequencies and Single Nucleotide Polymorphisms (SNPs).\n",
    "\n",
    "### Allele Frequencies\n",
    "\n",
    "Allele frequencies describe how often a particular variant (allele) of a gene appears within a given population. It is usually expressed as a proportion or percentage. Understanding allele frequencies is crucial for studying genetic diversity and for identifying alleles that may be associated with specific traits or diseases.\n",
    "\n",
    "### Single Nucleotide Polymorphisms (SNPs)\n",
    "\n",
    "SNPs are variations at a single position in a DNA sequence among individuals. They are the most common type of genetic variation and serve as markers for locating genes associated with diseases or specific traits.\n",
    "\n",
    "### What Makes a SNP a SNP?\n",
    "A Single Nucleotide Polymorphism occurs when a single nucleotide (A, T, C, or G) in the genome sequence is altered. For a variation to be considered a SNP, it must occur in at least 1% of the population. This distinguishes SNPs from random mutations, which are rare and may occur in any individual. SNPs can be synonymous (do not change the protein sequence) or non-synonymous (change the protein sequence), and they can occur in coding or non-coding regions of the genome.\n",
    "\n",
    "### Assumptions in SNP Analysis\n",
    "In SNP analysis, it is generally assumed that the nucleotide sequences between the SNPs are conserved, or identical, across the individuals being studied. This assumption allows us to focus on the SNPs as markers of genetic variation without having to analyze the entire genome.\n",
    "\n",
    "This is a reasonable assumption because:\n",
    "\n",
    "Most of the human genome is highly conserved.\n",
    "SNPs are the most common form of genetic variation, making them effective markers for genetic diversity.\n",
    "By focusing on SNPs, we can efficiently study genetic variation and its implications for traits, diseases, and population history.\n",
    "\n",
    "### Relatedness and Shared SNPs\n",
    "\n",
    "#### What Does It Mean If Two Individuals Share the Same SNP?\n",
    "\n",
    "Sharing a single SNP between two individuals doesn't necessarily imply a close genetic relationship, as SNPs can be quite common in populations. However, sharing the same SNPs over a specific length of DNA sequence can be a strong indicator of relatedness. **The length of the DNA sequence where the SNPs are shared is crucial.** A longer stretch of shared SNPs increases the likelihood that the two individuals are closely related. This is often measured in centimorgans (cM), a unit that describes the genetic distance between positions on a chromosome.\n",
    "\n",
    "#### Implications for Relatedness\n",
    "\n",
    "1. **Close Relatives**: Close relatives like siblings or parent-child pairs will share long stretches of SNPs.\n",
    "  \n",
    "2. **Distant Relatives**: More distant relatives like cousins may share shorter stretches but still longer than what would be expected by random chance.\n",
    "\n",
    "3. **Very Distantly Related Individuals**: In individuals who share a distant common ancestry, any shared SNPs are likely to be short, often less than 1 centimorgan (cM). These short stretches are scattered randomly across the genome. While they are not indicative of a close or recent familial relationship, they do reflect shared ancestors from a distant past.\n",
    "\n",
    "By analyzing the length and distribution of shared SNPs, researchers can infer the degree of relatedness between individuals, which is valuable in studies including Migration Patterns, Adaptation Studies, Archaeogenetics, Forensic Anthropology, and Evolutionary Anthropology.\n",
    "\n",
    "### Identifying SNPs\n",
    "\n",
    "1. **Total SNPs**: The total number of SNPs identified in the dataset gives us an overview of the genetic variation present.\n",
    "  \n",
    "2. **SNPs by Superpopulation**: Breaking down the SNPs by superpopulation allows us to understand how genetic variations are distributed among different groups. This is crucial for studies that aim to understand the genetic basis of population-specific traits or susceptibilities."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of samples: 500\n",
      "Total number of variants: 1,817,492\n",
      "Total number of biallelic variants: 1,706,442\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "gt = allel.GenotypeArray(callset['calldata/GT'])\n",
    "\n",
    "# Get the reference and alternate alleles from the VCF file.\n",
    "ref_alleles = callset['variants/REF']\n",
    "alt_alleles = callset['variants/ALT'][:, 0] #  to extract the first alternate allele for each variant. \n",
    "# This slice operation is taking all rows (:) and the first column (0) of the alt_alleles array\n",
    "\n",
    "# Create a boolean mask for biallelic variants\n",
    "is_biallelic = np.array([len(set(allele) - {'', '.'}) == 1 for allele in alt_alleles])\n",
    "\n",
    "# Create a boolean mask for valid SNPs\n",
    "valid_bases = {'A', 'T', 'C', 'G'}\n",
    "is_valid_snp = np.array([ref in valid_bases and alt in valid_bases for ref, alt in zip(ref_alleles, alt_alleles)])\n",
    "\n",
    "# Combine the two masks\n",
    "final_mask = is_biallelic & is_valid_snp\n",
    "\n",
    "# Apply the mask to the genotype array\n",
    "gt_filtered = gt.compress(final_mask, axis=0)\n",
    "\n",
    "print(f\"Total number of samples: {gt.shape[1]}\")\n",
    "print(f\"Total number of variants: {gt.shape[0]:,}\")\n",
    "print(f\"Total number of biallelic variants: {gt_filtered.shape[0]:,}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Understanding the Shape of the Genotype Array\n",
    "\n",
    "In the code, `gt` is a `GenotypeArray` object that contains the genotype data. The `shape` attribute of this array provides us with the dimensions of the array. Specifically, `gt.shape` returns a tuple with the following information:\n",
    "\n",
    "- `gt.shape[0]`: This represents the total number of variants (or genetic markers like SNPs) in the dataset.\n",
    "- `gt.shape[1]`: This represents the total number of samples (or individuals) in the dataset.\n",
    "\n",
    "By printing these values, we get a quick overview of the dataset's size, both in terms of the number of variants it contains and the number of samples it pertains to. This is valuable for understanding the scope of the data we are working with.\n",
    "\n",
    "### What is an Array?\n",
    "\n",
    "An array is a data structure that stores a collection of items, typically of the same type, at contiguous memory locations. The elements can be accessed randomly by indexing into the array. In programming languages like Python, arrays are often used for data manipulation and storage.\n",
    "\n",
    "### Understanding `gt.shape` in scikit-allel's GenotypeArray\n",
    "\n",
    "The `gt.shape` attribute in the context of a `GenotypeArray` from scikit-allel returns a tuple with three values, each representing a specific dimension of the genetic data:\n",
    "\n",
    "1. **`n_variants`**: The first value `gt.shape[0]` represents the number of genetic variants in the dataset, such as SNPs and indels.\n",
    "  \n",
    "2. **`n_samples`**: The second value `gt.shape[1]` represents the number of individual samples in the dataset.\n",
    "\n",
    "3. **`ploidy`**: The third value `gt.shape[2]` represents the ploidy of the organism, which is the number of sets of chromosomes in a cell. In simpler terms, it's the number of alleles for each variant.\n",
    "\n",
    "#### Why Use Arrays?\n",
    "\n",
    "Arrays are particularly useful when you need to store and manipulate large amounts of data efficiently. They are often used in mathematical computations, data analysis, and algorithmic implementations because of their efficiency in random access and data manipulation.\n",
    "\n",
    "In the context of genetics and bioinformatics, arrays can be used to store large datasets like genotype information, sequence data, and much more, allowing for efficient computations and analyses."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculate allele counts\n",
    "\n",
    "We use our gt variable that we created to calculate the allele counts using the count_alleles() method. \n",
    "This gives us the number of occurrences of each allele for each variant.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1000    0\n",
      "1000    0\n",
      "1000    0\n",
      "1000    0\n",
      " 999    1\n",
      "\n",
      "In many cases, all 500 individuals (1,000 haplotpes) have the same variant. However, there is some variability.\n",
      "Total number of genetic variants that vary: 878,069\n"
     ]
    }
   ],
   "source": [
    "ac = gt_filtered.count_alleles()\n",
    "print(ac[:5]) # print the first 5 variants\n",
    "\n",
    "max_allele_count = ac.max(axis=1)\n",
    "total_allele_count = ac.sum(axis=1)\n",
    "\n",
    "does_vary = (total_allele_count - max_allele_count) > 0\n",
    "# does_vary = (ac.sum(axis=1) - ac.max(axis=1)) > 0\n",
    "print(f\"In many cases, all 500 individuals (1,000 haplotpes) have the same variant. However, there is some variability.\")\n",
    "print(f\"Total number of genetic variants that vary: {np.count_nonzero(does_vary):,}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**What is ac?**\n",
    "\n",
    "`ac` is an Allele Counts Array, where each row represents a variant (like a SNP or an indel), and each column represents an allele. \n",
    "\n",
    "For bi-allelic SNPs, you'll typically have two columns: one for each allele (e.g., A and T).\n",
    "\n",
    "Here's a simplified example of what ac might look like for 3 variants:\n",
    "\n",
    "```\n",
    "20  30  # Variant 1: 20 counts of allele 1, 30 counts of allele 2\n",
    "40  10  # Variant 2: 40 counts of allele 1, 10 counts of allele 2\n",
    "25  25  # Variant 3: 25 counts of allele 1, 25 counts of allele 2\n",
    "```\n",
    "**What does axis=1 mean?**\n",
    "\n",
    "In NumPy (and many other Python libraries), an array can have multiple dimensions. The axis parameter specifies which dimension you want to perform the operation along.\n",
    "\n",
    "`axis=0` means that the operation will be performed vertically (down the rows for each column).\n",
    "`axis=1` means that the operation will be performed horizontally (across the columns for each row).\n",
    "\n",
    "For our example ac array:\n",
    "\n",
    "The sum for Variant 1 would be `20 + 30 = 50`\n",
    "\n",
    "The sum for Variant 2 would be `40 + 10 = 50`\n",
    "\n",
    "The sum for Variant 3 would be `25 + 25 = 50`\n",
    "\n",
    "So, ac.sum(axis=1) would return `[50, 50, 50]` in this example.\n",
    "\n",
    "**Why is this useful?**\n",
    "\n",
    "Summing the allele counts for each variant gives you the total number of alleles observed for that variant. This is useful for various types of genetic analyses, including identifying SNPs, calculating allele frequencies, and more.\n",
    "\n",
    "`ac.max(axis=1)`\n",
    "\n",
    "This finds the maximum allele count for each variant. Essentially, it identifies the most common allele for each variant.\n",
    "\n",
    "`(ac.sum(axis=1) - ac.max(axis=1)) > 0`\n",
    "\n",
    "This expression calculates the difference between the total number of alleles and the count of the most common allele for each variant. The idea is to find out how many of the \"other\" alleles are present.\n",
    "\n",
    "For example, if a variant has allele counts `[20, 30]`, the sum would be `50`, and the max would be `30`. The difference `50 - 30` would be `20`, representing the count of the less common allele.\n",
    "\n",
    "If the value of `ac.max(axis=1)` is the same as `ac.sum(axis=1)`, it means that all observed alleles for that particular variant are the same. In other words, there is only one type of allele present for that variant.\n",
    "\n",
    "For example, let's say a variant has allele counts `[50, 0]`. The sum would be `50`, and the max would also be `50`. The difference `50 - 50` would be `0`, and `(ac.sum(axis=1) - ac.max(axis=1)) > 0` would evaluate to `False` for this variant, indicating that it is not a SNP.\n",
    "\n",
    "So, when the sum and the max are the same, the variant is not considered a Single Nucleotide Polymorphism (SNP) because there is no variation; all observed alleles are the same.\n",
    "\n",
    "If the count of the less common allele is greater than zero, that means there is more than one type of allele present for that variant, making it a Single Nucleotide Polymorphism (SNP).\n",
    "\n",
    "`np.count_nonzero(is_snp)`\n",
    "Finally, this counts the number of `True` values in the `is_snp` array, giving you the total number of SNPs identified.\n",
    "\n",
    "Putting it all together\n",
    "The line `is_snp = (ac.sum(axis=1) - ac.max(axis=1)) > 0` is a way to identify SNPs by checking if there is more than one type of allele present for each variant. Then, np.count_nonzero(is_snp) counts how many SNPs have been identified."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Minor Allele Frequency\n",
    "What is Minor Allele Frequency?\n",
    "\n",
    "In a given population, for a particular genetic variant (or SNP), the allele that occurs less frequently is termed the \"minor allele.\" The frequency of this minor allele in the population is known as the Minor Allele Frequency (MAF). It is calculated as the count of the minor allele divided by the total number of alleles examined.\n",
    "\n",
    "**Why is MAF Important?**\n",
    "\n",
    "Statistical Power: Variants with extremely low MAF are often excluded because they may lack the statistical power to detect association with a trait or disease.\n",
    "\n",
    "Quality Control: Filtering by MAF is a common quality control step to remove potential errors in variant calling.\n",
    "\n",
    "Biological Relevance: Variants with higher MAF are more likely to be biologically relevant and less likely to be random mutations.\n",
    "\n",
    "Calculating MAF with scikit-allel\n",
    "In scikit-allel, you can calculate the MAF as follows:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Of the 1,706,442 variants in chromosome 20, we determine that 878,069 vary, but are not necessarly SNPs.\n"
     ]
    }
   ],
   "source": [
    "print(f\"Of the {gt_filtered.shape[0]:,} variants in chromosome 20, we determine that {np.count_nonzero(does_vary):,} vary, but are not necessarly SNPs.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of SNPs accross the whole dataset: 280,203\n",
      "Total number of common SNPs accross the whole dataset: 158,391\n"
     ]
    }
   ],
   "source": [
    "# Calculate Minor Allele Frequency (MAF)\n",
    "\n",
    "# max_allele_count = ac.max(axis=1)\n",
    "# total_allele_count = ac.sum(axis=1)\n",
    "\n",
    "minor_allele_count = total_allele_count - max_allele_count\n",
    "\n",
    "maf = (total_allele_count - max_allele_count) / total_allele_count\n",
    "\n",
    "\n",
    "# Here, gt.compress(maf > 0.05, axis=0) will keep only the rows (variants) in the genotype array where the MAF is greater than 5%.\n",
    "# A set of SNPs with a MAF greater than 5%\n",
    "gt_maf_filtered_5 = gt_filtered.compress(maf > 0.05, axis=0)\n",
    "print(f\"Total number of common SNPs accross the whole dataset: {gt_maf_filtered_5.shape[0]:,}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Summary\n",
    "Filtering by MAF is a crucial step in many genetic analyses. It helps in focusing on variants that are likely to be meaningful, thereby increasing the robustness and reliability of your results."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
